### PR TITLE
Fixing compilation if HPX_WITH_DATAPAR_VC is enabled

### PR DESCRIPTION
**Pull request by hkaiser**
- flyby: move datapar algorithm tests into algorithms module (from execution module)
- flyby: fix SFINAE conditions for locality constructor

### CODE REVIEW

File: hpx/runtime/parcelset/locality.hpp [hunk 1]
```diff
@@ -68,22 +68,22 @@ namespace hpx { namespace parcelset
...
-        locality(Impl && i)
-          : impl_(new impl<typename util::decay<Impl>::type>(std::forward<Impl>(i)))
+        template <typename Impl,
+            typename Enable1 = typename std::enable_if<!std::is_same<locality,
+                typename util::decay<Impl>::type>::value>::type,
+            typename Enable2 = typename std::enable_if<
+                !traits::is_iterator<Impl>::value>::type>
+        locality(Impl&& i)
+          : impl_(new impl<typename util::decay<Impl>::type>(
+                std::forward<Impl>(i)))
```
Comment (msimberg) [inline hpx/runtime/parcelset/locality.hpp]: What was the problem here btw?

Comment (hkaiser) [reply hpx/runtime/parcelset/locality.hpp]: Heh, the actual problem was this: https://github.com/STEllAR-GROUP/hpx/pull/4129/files#diff-f63ab7a8005a210ebd446a9e06850ba9R55. Sorry for the flyby changes.

Comment (msimberg) [reply hpx/runtime/parcelset/locality.hpp]: No worries! I was wondering if this particular change fixed something that didn't work. Or does this save us some template instantiations or something similar?

Comment (hkaiser) [reply hpx/runtime/parcelset/locality.hpp]: The old code started to break for me after I enabled the datapar stuff. Not sure why it wasn't broken before, but it's a similar issue as we fixed yesterday for the `unique_any`.

Comment (msimberg) [reply hpx/runtime/parcelset/locality.hpp]: All right, thanks!

**Pull request merged**